### PR TITLE
Parse the DelCall type, update the typescript version, use ts-jest and more...

### PR DESCRIPTION
**Pull request by matuzalemsteles**
hey guys, I did some index signature fixes that was preventing from compiling and using elsewhere (e.g [generate-soy-source-maps](https://github.com/matuzalemsteles/generate-soy-source-maps)) and added `DelCall` type support.

> It would be interesting if we followed the AST of the soy parser in the closure template, it would have a lot of breaking change, but it would help to have a more complete analysis of the use cases. See [SoyFileParser.jj](https://github.com/google/closure-templates/blob/master/java/src/com/google/template/soy/soyparse/SoyFileParser.jj)

### CODE REVIEW

File: test/fixtures/DelCall.soy [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+{deltemplate Test.First variant="'foo'"}
+	{@param name: string}
+	{@param title: string}
+
+	<div>
+		<h1>{$title}</h1>
+		<h2>{$name}</h2>
+	</div>
+{/deltemplate}
```
Comment (jbalsas) [inline test/fixtures/DelCall.soy]: Could we add some more cases?

```soy
{delcall Test.Empty variant="foo" /}
{delcall Test.Data variant="foo" data="all" /}
{delcall Test.EmptyDefault allowemptydefault="true" /}
...
```

**Review by jbalsas — CHANGES_REQUESTED** [review]
Hey @matuzalemsteles, this looks very promising!! 👍 

Would you mind adding some additional cases, just to make sure we're covering other usages of `delcall` in the parser?

Also... could you please commit a `package-lock` instead? 🙏 

@mthadley, wouldn't mind if you took a look here 😉

**Comment by mthadley** [comment]
@jbalsas Sure, no problemo.

@matuzalemsteles I agree that it would have been nicer at this point to have started by following the SoyParser grammar as close as possible. However, this module evolved out of what started as a simple script, so the planning just wasn't there.

Maybe in the future though, someone can work on that.

File: src/parser.ts [hunk 2]
```diff
@@ -133,6 +133,19 @@ const call = nodeMap(
       .skip(spaced(closeCmd('call'))))
 );
 
+const delCall = nodeMap(
+  S.DelCall,
+  P.string('{delcall')
+    .skip(P.whitespace)
+    .then(templateName),
+  optional(P.seq(P.whitespace, P.string('variant='))
```
Comment (mthadley) [inline src/parser.ts]: I think we can use the `attribute` parser here, and maybe combine it with `many` to handle the cases that @jbalsas mentioned.

File: src/traverse.ts [hunk 3]
```diff
@@ -27,7 +28,7 @@ export interface Visitor {
   StringLiteral?: Visit<S.StringLiteral>;
   Template?: Visit<S.Template>;
   Ternary?: Visit<S.Ternary>;
-  [propName: string]: Visit<S.Node> | undefined;
+  [propName: string]: Visit<any> | undefined;
```
Comment (mthadley) [inline src/traverse.ts]: If at all possible, I'd like to avoid having to make this change. I'm not sure what compiler error you are running into, but maybe we can find a different way to fix it? If anything, I think I'd rather find a way to remove this index signature.

Replacing this with `any` is means we lose type safety here. I don't remember exactly why I added it, but I think this index signature was more of a fallback. All of the other `Visitor` attributes implement `Node`, so this unified with those, at least at a more general level.

TLDR: Let's try to keep this as `Visit<S.Node>`, but if we absolutely have no other options, then maybe we can look at changing the type (like replacing it with `unknown`, `any`, etc.).

**Review by mthadley — CHANGES_REQUESTED** [review]
@matuzalemsteles Nice work. It's really cool to see someone working on source maps. Interested to see how it turns out. 💯

**Comment by jbalsas** [comment]
Thanks a lot @mthadley! I'd love to follow the official (yet internal) Soy grammar. We need to keep pushing to improve developer experience bit by bit, so let's see if we can find time to work on it! 👍

**Comment by matuzalemsteles** [comment]
@jbalsas Thanks. Yes, I'll add more cases. I had forgotten these, thank you for reminding me.

@mthadley Thanks. I understand, I realized this when I looked at some old commits and issues... Feel free to contribute to the source map 🙂.

I intend to improve it and to follow with the Soy grammar in the future... With this, we can have greater freedom with the soy. (Maybe even use it in our own compiler of soy.... just thoughts 🙂)

File: src/parser.ts (hunk 2 above)
Comment (matuzalemsteles) [reply src/parser.ts]: Thanks @mthadley. I'm getting acquainted with the parsimmon.

File: src/traverse.ts (hunk 3 above)
Comment (matuzalemsteles) [reply src/traverse.ts]: hey @mthadley, I was also thinking of that, as everyone else also extends from `Node`, would have no problems. But when it extends seems to lose its reference...

But there is an error when trying to compile:
```typescript
[ts] Property 'BooleanLiteral' of type 'VisitFunction<BooleanLiteral> | VisitObject<BooleanLiteral> | undefined' is not assignable to string index type 'VisitFunction<Node> | VisitObject<Node> | undefined'.
(property) Visitor.BooleanLiteral?: VisitFunction<S.BooleanLiteral> | VisitObject<S.BooleanLiteral> | undefined
```

This happens in all `Visitor` properties. We could solve it this way:
```typescript
export type Visit<T> = VisitFunction<keyof T> | VisitObject<keyof T>;
```

But we will be just taking the properties of each interface... I'm thinking of other ways how we can handle this. Do you have any suggestions for not going with `any`?

**Comment by mthadley** [comment]
@matuzalemsteles 

So I took a look at your problem, and I think I found a better way to represent the `Vistor` type:

https://github.com/mthadley/soyparser/commit/adc7ae7f24728b2a4b7f9cf70a696ffb0154b229

It looks like we still have type safety for the visitor, and we actually gained some. These changes also caused me to find a typo that had been there for years now, `NumberLital` should have been `NumberLiteral` in the visitor definition. 😆 

If you can, take a look to see if that makes sense. Feel free to incorporate it into this PR if it does.

**Note**: This actually introduces a breaking change since I changed the way that `OtherCmd` was represented in the AST. I don't think it will matter though. I'm not aware of any code that was relying on the old behavior. It was just there as a catch all.

**Comment by matuzalemsteles** [comment]
@mthadley 

Thanks so much for your thoughts, I think we can go with this and go modeling AST to get closer to Soy Grammar so we will remove `OtherCmd` anytime... I tried to see other solutions but I did not get very close. I'll carry on with it.

**Comment by mthadley** [comment]
@matuzalemsteles Sounds great. I only created that commit to avoid having to introduce `any` in the `Visitor`. If you have any problems with it let me know, otherwise, everything else looks great!

**Review by mthadley — APPROVED** [review]

**Review by jbalsas — APPROVED** [review]

**Comment by jbalsas** [comment]
Hey @mthadley, I merged this, but I'm having some problems to publish a new `1.1.0` version... likely because of some incorrect 2FA configuration...

Could you please take a look?

**Comment by mthadley** [comment]
@matuzalemsteles @jbalsas Published as [`v2.0.0`](https://github.com/metal/soyparser/releases/tag/v2.0.0).

**Comment by matuzalemsteles** [comment]
Thanks @mthadley! 🎉
